src/Compatibility/Microsoft.DotNet.ApiCompatibility/Rules/CannotRemoveBaseTypeOrInterface.cs

```diff
@@ -0,0 +1,85 @@
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.DotNet.ApiCompatibility.Abstractions;
+using Microsoft.DotNet.ApiCompatibility.Extensions;
+
+namespace Microsoft.DotNet.ApiCompatibility.Rules
+{
+    public class CannotRemoveBaseTypeOrInterface : Rule
+    {
+        public override void Initialize(RuleRunnerContext context)
+        {
+            context.RegisterOnTypeSymbolAction(RunOnTypeSymbol);
+        }
+
+        private void RunOnTypeSymbol(ITypeSymbol left, ITypeSymbol right, string leftName, string rightName, IList<CompatDifference> differences)
+        {
+            if (left == null || right == null)
+                return;
+
+            if (left.TypeKind != TypeKind.Interface && right.TypeKind != TypeKind.Interface)
+            {
+                // if left and right are not interfaces check base types
+                ValidateBaseTypeNotRemoved(left, right, leftName, rightName, differences);
+
+                if (Settings.StrictMode)
+                    ValidateBaseTypeNotRemoved(right, left, rightName, leftName, differences);
+            }
+
+            ValidateInterfaceNotRemoved(left, right, leftName, rightName, differences);
+
+            if (Settings.StrictMode)
+                ValidateInterfaceNotRemoved(right, left, rightName, leftName, differences);
+        }
+        private void ValidateBaseTypeNotRemoved(ITypeSymbol left, ITypeSymbol right, string leftName, string rightName, IList<CompatDifference> differences)
+        {
+            // Base type order matters in the hierarchy chain so therefore use a list.
+            List<ITypeSymbol> rightBaseTypes = right.GetAllBaseTypes().ToList();
+
+            int lastTypeIndex = 0;
+            foreach (ITypeSymbol leftBaseType in left.GetAllBaseTypes())
+            {
+                lastTypeIndex = rightBaseTypes.FindIndex(lastTypeIndex, r => Settings.SymbolComparer.Equals(leftBaseType, r));
+
+                if (lastTypeIndex < 0)
+                {
+                    differences.Add(new CompatDifference(
+                                            DiagnosticIds.CannotRemoveBaseType,
+                                            string.Format(Resources.CannotRemoveBaseType, left.ToDisplayString(), leftBaseType.ToDisplayString(), rightName, leftName),
+                                            DifferenceType.Changed,
+                                            right));
+                    return;
+                }
+            }
+        }
+
+        private void ValidateInterfaceNotRemoved(ITypeSymbol left, ITypeSymbol right, string leftName, string rightName, IList<CompatDifference> differences)
+        {
+            HashSet<ITypeSymbol> rightInterfaces = new(right.GetAllBaseInterfaces(), Settings.SymbolComparer);
+
+            foreach (ITypeSymbol leftInterface in left.GetAllBaseInterfaces())
+            {
+                // Ignore internal interfaces
+                if (!leftInterface.IsVisibleOutsideOfAssembly())
```

Comment:
should we add a flag here if ```Includeinternals``` is set.

Reply:
I thought about it, however removing an internal interface is not a breaking change. Maybe if we do that we should do the same for removing base types?

cc: @dougbu @ericstj thoughts?

Reply:
we hit this one for old api compat for open telemetry

Reply:
Yeah I think you are right we should consider that flag as if internals are included that means a lot of types and interfaces  that are internal will only be able to implement internal interfaces. I don't know if we should for the base type scenario though, don't know what the policy is there.

Reply:
I added a change to include internals if the flag is set, however abstract classes check is always including non visible outside of the assembly, I don't know if we should filter that out. Don't know if removing an internal or private class from a hierarchy is breaking (my assumption is it is as it could inherit from public base types and we would loose those in the hierarchy?).